contrib/cf-diag.sh

```diff
@@ -0,0 +1,240 @@
+
+# Note also THIS TEST MAY FAIL ON NON-LINUX SYSTEMS.
+# There is no portable approach I can find for this 'stat' command.
+# However, I consider this a minor (negligible) issue.
+# (It will also fail with coreutils 8.6, but 8.5 and 8.7+ are fine:
+# see https://unix.stackexchange.com/q/3348/135943.)  :)
+
+desc="cf-agent ran in last 15 minutes"
+if [ "$(($(date +%s)-$(stat -c %Y /var/cfengine/promise_summary.log)))" -lt 900 ]; then
+  succeed "$desc"
+else
+  fail "$desc"
+fi
+
+# HUB CHECKS
+
+desc="Bootstrapped to hub (policy_server.dat exists)"
+if [ -f /var/cfengine/policy_server.dat ]; then
+  succeed "$desc"
+  hub_ip="$(cat /var/cfengine/policy_server.dat)"
+  printf '# Hub IP is %s\n' "$hub_ip"
+else
+  fail "$desc"
+  bail 'Not connected to any hub.'
+fi
+
+desc="Hub is connecting to host"
+if cf-key -s | grep ^Incoming | grep -qF "$hub_ip"; then
+  succeed "$desc"
+
+  hub_lastseen="$(cf-key -s |grep -F "$hub_ip" |awk '/^Incoming/{print $4, $5, $6, $7, $8}')"
+  hub_lastseen_in_seconds="$(date +%s -d "$hub_lastseen")"
+  now="$(date +%s)"
+  seconds_since_conn="$((now-hub_lastseen_in_seconds))"
+
+  desc="Hub connected to host within last 15 minutes"
+  if [ "$seconds_since_conn" -lt 900 ]; then
+    succeed "$desc"
+  else
+    fail "$desc"
+  fi
+  printf '# Has been %d seconds since last connection\n' "$seconds_since_conn"
+
+else
+  # Hub connection fail
+  fail "$desc"
+fi
+
+### SYSTEM CHECKS
+
+disk_usage="$(df -P / | awk '/\//{print $5}')"
+
+desc="Filesystem has free space"
+if [ "$disk_usage" = '100%' ]; then
+  fail "$desc"
+  bail "Can't continue with full disk"
+else
+  succeed "$desc"
+  printf '# Disk usage is %s\n' "$disk_usage"
+fi
+
+desc="Filesystem is writable"
+if touch /tmp/cf_diag_test 2>/dev/null; then
+  succeed "$desc"
+  rm /tmp/cf_diag_test
+else
+  fail "$desc"
+  bail "Can't continue with read-only filesystem"
+fi
+
+### POLICY CHECKS
+
+for f in promises.cf update.cf; do
+  desc="$f exists"
+  if [ -f /var/cfengine/inputs/"$f" ]; then
+    succeed "$desc"
+  else
+    fail "$desc"
+    bail "No $f"
+  fi
+done
+
+# The tests below this point may take a long time to run,
+# depending on how efficiently your CFEngine policy is written.
+
+for f in promises.cf update.cf; do
+  desc="$f is runnable"
+  if cf-agent -f "$f" >/dev/null 2>&1; then
+    succeed "$desc"
+  else
+    fail "$desc"
+    bail "$f not runnable"
+  fi
+done
+
+desc="Update run fixes changes to policy"
+# Running update policy (after deleting flag file)
+# should repair alteration of promises.cf
+echo '#cf-diag_comment' >> /var/cfengine/inputs/promises.cf
+rm -f /var/cfengine/inputs/cf_promises_validated
+cf-agent -K -f update.cf >/dev/null 2>&1
+if grep -q cf-diag_comment /var/cfengine/inputs/promises.cf; then
+  fail "$desc"
+else
+  succeed "$desc"
+fi
+
+############################### CUSTOM TESTS ##################################
+
+# Add custom tests here (if you really need them).
+# Use the general form:
```

Comment:
Humm, seems like a good place to be looking for a directory full of tests matching some pattern.

Splitting the tests out into directories full of a kind of test might be intersting as well.
Then you could add notes about adding a `.gitlab-ci.yml` spec which runs some of the scripts as tests on your policy repo.

Reply:
I did consider that.  But really, the purpose of this script is emphatically NOT to be a general-purpose testing framework.  Customized tests are not the point.  A correctly sequenced battery of standard universally applicable (or close to that) tests, is the point.

If I wanted to write a whole shell scripting testing framework, I would have just reinvented BATS.  But that's not the point of this script.  This script is intended to be ultra-lightweight, simple, standalone, and to work even on very old servers (e.g. RHEL 4).  And is intended to give a quick overview of the CFEngine health factors on the system.  And really, nothing beyond that.